singularity-eos/eos/eos_helmholtz.hpp

```diff
@@ -0,0 +1,1439 @@
+                     const Real lDe, Real *lambda) const;
+
+  static constexpr Real ROOT_THRESH = 1e-14;
+  Options options_;
+  HelmRad rad_;
+  HelmIon ions_;
+  HelmCoulomb coul_;
+  HelmElectrons electrons_;
+};
+
+PORTABLE_INLINE_FUNCTION
+void Helmholtz::FillEos(Real &rho, Real &temp, Real &energy, Real &press, Real &cv,
+                        Real &bmod, const unsigned long output, Real *lambda) const {
+  using namespace HelmUtils;
+  bool need_temp = (output & thermalqs::temperature);
+  bool need_sie = (output & thermalqs::specific_internal_energy);
+  bool need_rho = (output & thermalqs::density);
+  PORTABLE_REQUIRE(!need_rho, "Density output not supported by this EOS");
+  PORTABLE_REQUIRE(!(need_temp && need_sie),
+                   "Either specific internal energy or temperature must be provided.");
+  Real abar = lambda[Lambda::Abar];
+  Real zbar = lambda[Lambda::Zbar];
+  Real ytot, ye, ywot, De, lDe, lT;
+  GetElectronDensities_(rho, abar, zbar, ytot, ye, ywot, De, lDe);
+  if (need_temp) {
+    lT = lTFromRhoSie_(rho, energy, abar, zbar, ye, ytot, ywot, De, lDe, lambda);
+    temp = math_utils::pow10(lT);
+  } else {
+    lT = std::log10(temp);
+    lambda[Lambda::lT] = lT;
+  }
+  Real p[NDERIV], e[NDERIV], s[NDERIV], etaele[NDERIV], nep[NDERIV];
+  GetFromDensityLogTemperature_(rho, temp, abar, zbar, ye, ytot, ywot, De, lDe, p, e, s,
+                                etaele, nep);
+  if (output & thermalqs::specific_internal_energy) {
+    energy = e[0];
+  }
+  if (output & thermalqs::pressure) {
+    press = p[0];
+  }
+  if (output & thermalqs::specific_heat) {
+    cv = e[DDT];
+  }
+  if (output & thermalqs::bulk_modulus) {
+    Real gamma1 = ComputeGamma1_(rho, temp, p, e);
+
+    // Note this is the bulk modulus, not the sound speed. To compute
+    // the sound speed, you must take this value and divide by the
+    // enthalpy. In particular:
+    // c_s = sqrt(bmod / (w)) = sqrt(bmod / (h rho))
+    // see page 108 of Rezzolla and Zanotti
+    bmod = std::max(robust::EPS(), p[0] * gamma1);
+  }
+}
+
+PORTABLE_INLINE_FUNCTION
+Real Helmholtz::lTFromRhoSie_(const Real rho, const Real e, const Real abar,
+                              const Real zbar, const Real ye, const Real ytot,
+                              const Real ywot, const Real De, const Real lDe,
+                              Real *lambda) const {
+  using namespace HelmUtils;
+  const Real abari = robust::ratio(1.0, abar);
+  const Real ni = abari * rho * ions_.NA;
+  const Real ne = zbar * ni;
+  Real lT;
+  Real T;
+
+  if (options_.ENABLE_RAD || options_.GAS_DEGENERATE ||
+      options_.ENABLE_COULOMB_CORRECTIONS) {
+    Real lTguess = lambda[Lambda::lT];
+    if (!((electrons_.lTMin() <= lTguess) && (lTguess <= electrons_.lTMax()))) {
+      lTguess = lTAnalytic_(rho, e, ni, ne);
+      if (!((electrons_.lTMin() <= lTguess) && (lTguess <= electrons_.lTMax()))) {
+        // Fallback initial guess
+        lTguess = 7;
```

Comment:
This was meant to be a fallback initial guess in the middle of the table range (in case the analytic guess fails/ is outside the table range for some reason), I just failed to do a simple calculation. Fixed now to 8 (table range is from 3 to 13). It should in any case not matter that much since it is an arbitrary guess anyway.

Reply:
Tmin and tmax are defined as constants right? We could slightly future proof this by setting ltguess to explicitly be the average of ltmin and ltmax.